File: src/main/java/net/andrecarbajal/naviMusic/audio/spotify/SpotifyTokenProvider.java
package net.andrecarbajal.naviMusic.audio.spotify;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import se.michaelthelin.spotify.SpotifyApi;
import se.michaelthelin.spotify.model_objects.credentials.ClientCredentials;
import se.michaelthelin.spotify.requests.authorization.client_credentials.ClientCredentialsRequest;

@Slf4j
@Component
public class SpotifyTokenProvider {
    private final SpotifyApi spotifyApi;
    private String accessToken;
    private long tokenExpirationTime = 0;

    public SpotifyTokenProvider(@Value("${app.spotify.clientId}") String clientId,
                                @Value("${app.spotify.clientSecret}") String clientSecret) {
        this.spotifyApi = new SpotifyApi.Builder()
                .setClientId(clientId)
                .setClientSecret(clientSecret)
                .build();
        refreshAccessToken();
    }

    public synchronized String getAccessToken() {
        if (System.currentTimeMillis() >= tokenExpirationTime) {
            refreshAccessToken();
        }
        return accessToken;
    }

    public SpotifyApi getSpotifyApi() {
        spotifyApi.setAccessToken(getAccessToken());
        return spotifyApi;
    }

    private void refreshAccessToken() {
        try {
            ClientCredentialsRequest clientCredentialsRequest = spotifyApi.clientCredentials().build();
            ClientCredentials clientCredentials = clientCredentialsRequest.execute();
            this.accessToken = clientCredentials.getAccessToken();
            this.tokenExpirationTime = System.currentTimeMillis() + (clientCredentials.getExpiresIn() - 60) * 1000L; // 60s margen
            spotifyApi.setAccessToken(this.accessToken);
        } catch (Exception e) {
            log.error("Error refreshing Spotify access token: {}", e.getMessage());
        }
    }
}

File: src/main/kotlin/net/andrecarbajal/naviMusic/audio/spotify/SpotifyTokenProvider.kt
package net.andrecarbajal.naviMusic.audio.spotify

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import se.michaelthelin.spotify.SpotifyApi

@Component
class SpotifyTokenProvider(
    @Value($$"${app.spotify.clientId}") private val clientId: String,
    @Value($$"${app.spotify.clientSecret}") private val clientSecret: String
) {
    private val log = LoggerFactory.getLogger(SpotifyTokenProvider::class.java)

    private val spotifyApi: SpotifyApi = SpotifyApi.Builder()
        .setClientId(clientId)
        .setClientSecret(clientSecret)
        .build()

    private var accessToken: String? = null
    private var tokenExpirationTime: Long = 0

    init {
        if (isConfigured()) {
            refreshAccessToken()
        } else {
            log.warn("Spotify API is not configured. Spotify links will not work.")
        }
    }

    fun isConfigured(): Boolean {
        return clientId.isNotBlank() && clientId != "default" && 
               clientSecret.isNotBlank() && clientSecret != "default"
    }

    @Synchronized
    fun getAccessToken(): String? {
        if (!isConfigured()) return null

        if (System.currentTimeMillis() >= tokenExpirationTime) {
            refreshAccessToken()
        }
        return accessToken
    }

    fun getSpotifyApi(): SpotifyApi {
        spotifyApi.accessToken = getAccessToken()
        return spotifyApi
    }

    private fun refreshAccessToken() {
        if (!isConfigured()) return
        try {
            val clientCredentialsRequest = spotifyApi.clientCredentials().build()
            val clientCredentials = clientCredentialsRequest.execute()
            this.accessToken = clientCredentials.accessToken
            this.tokenExpirationTime = System.currentTimeMillis() + (clientCredentials.expiresIn - 60) * 1000L
            spotifyApi.accessToken = this.accessToken
        } catch (e: Exception) {
            log.error("Error refreshing Spotify access token: {}", e.message)
        }
    }
}
